tests/components/eafm/test_sensor.py

```diff
@@ -0,0 +1,431 @@
+    """
+    Test that a measure can be a dict not a list.
+
+    E.g. https://environment.data.gov.uk/flood-monitoring/id/stations/751110
+    """
+    _ = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": {
+                "@id": "really-long-unique-id",
+                "label": "York Viking Recorder - level-stage-i-15_min----",
+                "qualifier": "Stage",
+                "parameterName": "Water Level",
+                "latestReading": {"value": 5},
+                "stationReference": "L1234",
+            },
+        },
+    )
+
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "5"
+
+
+async def test_reading_no_unit(hass, mock_get_station):
+    """
+    Test that a sensor functions even if its unit is not known.
+
+    E.g. https://environment.data.gov.uk/flood-monitoring/id/stations/L0410
+    """
+    _ = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": [
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Water Level",
+                    "latestReading": {"value": 5},
+                    "stationReference": "L1234",
+                }
+            ],
+        },
+    )
+
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state.state == "5"
+
+
+async def test_ignore_invalid_latest_reading(hass, mock_get_station):
+    """
+    Test that a sensor functions even if its unit is not known.
+
+    E.g. https://environment.data.gov.uk/flood-monitoring/id/stations/L0410
+    """
+    _ = await async_setup_test_fixture(
+        hass,
+        mock_get_station,
+        {
+            "label": "My station",
+            "measures": [
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Water Level",
+                    "latestReading": "http://environment.data.gov.uk/flood-monitoring/data/readings/L0410-level-stage-i-15_min----/2017-02-22T10-30-00Z",
+                    "stationReference": "L0410",
+                },
+                {
+                    "@id": "really-long-unique-id",
+                    "label": "York Viking Recorder - level-stage-i-15_min----",
+                    "qualifier": "Stage",
+                    "parameterName": "Other",
+                    "latestReading": {"value": 5},
+                    "stationReference": "L0411",
+                },
+            ],
+        },
+    )
+
+    state = hass.states.get("sensor.my_station_water_level_stage")
+    assert state is None
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)